odb_extractor_atoms.py

```diff
@@ -0,0 +1,627 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Code for manipulating ODB Extractor json files and initial sequence atom definitions
+# Bryan Miller
+# 2021-11-24
+
+import os
+import json
+import gzip
+
+from openpyxl import Workbook
+from openpyxl import load_workbook
+
+
+fpuinst = {'GNIRS': 'instrument:slitWidth', 'GMOS-N': 'instrument:fpu'}
+
+
+def findatoms(sequence):
+    
+    qoffsets = []
+    atoms = []
+    natom = 0
+    nabba = 0
+    
+    nsteps = len(sequence)
+    
+    for ii, step in enumerate(sequence):
+        nextatom = False
+        keys = list(step.keys())
+
+        datalab = step['observe:dataLabel']
+        observe_class = step['observe:class']
+        exptime = float(step['observe:exposureTime'])
+        inst = step['instrument:instrument']
+    #     print(inst, fpuinst[inst])
+        fpu = step[fpuinst[inst]]
+        if 'instrument:filter' in keys:
+            filter_name = step["instrument:filter"]
+        else:
+            filter_name = 'None'
+        wavelength = float(step['instrument:observingWavelength'])
+        step_time = step['totalTime']/1000.
+        
+        if 'GMOS' in inst:
+            coadds = 1
+        else:
+            coadds = int(step['observe:coadds'])
+
+        disperser = step['instrument:disperser']
+
+        # Offsets
+        if 'telescope:p' in keys:
+            p = float(step['telescope:p'])
+        else:
+            p = 0.0
```

Comment:
Just curious why p offsets are not held in a list, but q offsets are?

Reply:
I was only only concerned with identifying ABBA patterns which just offset in q. The qoffsets list was an early idea and is not actually used. In my current version I use the qoffsets variable for something else.